include/dali/core/access_order.h

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_ACCESS_ORDER_H_
+#define DALI_CORE_ACCESS_ORDER_H_
+
+#include <cstddef>
+#include "dali/core/api_helper.h"
+#include "dali/core/cuda_error.h"
+#include "dali/core/cuda_stream.h"
+
+namespace dali {
+
+/**
+ * @brief Describes the order in which an object is accessed. It can be stream or host.
+ *
+ * The access order can be either a CUDA stream order or host order. There's also
+ * a special "null" value, useful when passing AccessOrder to functions/objects which may have
+ * internal streams.
+ */
+class DLL_PUBLIC AccessOrder {
+ public:
+  constexpr AccessOrder() = default;
+  constexpr AccessOrder(cudaStream_t stream, int device_id)
+  : stream_(stream), device_id_(device_id) {}
+
+  AccessOrder(int) = delete;  // NOLINT  prevent construction from 0
+  AccessOrder(std::nullptr_t) = delete;  // NOLINT
+
+  constexpr AccessOrder(cudaStream_t stream) : stream_(stream) {  // NOLINT
+    if (is_device())
+      device_id_ = DeviceFromStream(stream);
+  }
+
+  static constexpr AccessOrder host() {
+    return AccessOrder(host_sync_stream());
+  }
+
+  cudaStream_t get() const noexcept { return has_value() ? stream_ : 0; }
+
+  cudaStream_t stream() const noexcept { return is_device() ? stream_ : 0; }
```

Comment:
I'm bit confused by two of this having similar but quite different behavior.

Reply:
The latter is mostly for interoperability with legacy functions like TypeInfo::Copy, which use a value of 0 in cases where the stream is not relevant and which cannot handle the `host_sync_stream()` value in a meaningful way.

Reply:
Can you add a comment and write that `get` should be used when possible?